node/core/backing/src/lib.rs

```diff
@@ -974,6 +976,42 @@ async fn handle_active_leaves_update<Context>(
 				},
 			);
 
+			// Check if any space opened up in a new fragment tree for pending requests.
+			for requests in state.seconding_check_requests.values_mut() {
+				for request in std::mem::take(requests) {
+					let request_params = request.request;
+					// Only perform seconding check for a single inserted leaf.
+					let result = seconding_sanity_check(
+						ctx,
+						state.per_leaf.get_key_value(&leaf.hash),
+						&state.implicit_view,
+						request_params.candidate_hash,
+						request_params.candidate_para_id,
+						request_params.parent_head_data_hash,
+						request_params.candidate_relay_parent,
+					)
+					.await;
+
+					match result {
+						SecondingAllowed::No => {
+							// unreachable: the request wasn't rejected previously
+							// when received, it's not expected to have any possible
+							// membership in known fragment trees. Once there appears
+							// at least one, request is removed from the memory.
+							// Thus, we couldn't have seconded candidate with the same
+							// fragment tree membership prior to this activation.
+						},
+						SecondingAllowed::Yes(membership) =>
+							if membership.iter().any(|(_, m)| !m.is_empty()) {
+								let _ = request.tx.send(true);
```

Comment:
This still looks incorrect. We only want to send `true` when it's allowed in all leaves, not just in one leaf, right?

Reply:
Seconding sanity check would return early with `No` if it's disallowed somewhere
Otherwise, I don't see a reason to stop collators from building on a single fork if they want to (for example)

Reply:
The goal of the seconding check isn't to prevent collators from doing anything, FWIW. It's to prevent validators from misbehaving.

The collator-protocol should also retry these requests when it gets new leaves, if the expectation is for them to fail fast.

But as far as I can tell, the logic is like this:
  1. We fail fast when we receive the request if there is a conflicting entry at any existing leaf
  1. Therefore at the point where this runs the first time, we don't have any conflicting entries - rather, we just don't recognize fragment tree membership
  1. In the meantime, we may learn of new backed candidates, which either make fragment tree membership clear, but if so we already returned (true), so this code won't run later.
  1. We are importing a single new leaf, which may make fragment tree membership clear
  1. Either it has a possible membership in the fragment tree (we return true) or it does not (precondition at (2) holds)

Reply:
OK I didn't get your comment first, this question is reasonable (thought it points to a different part of code).

And your explanation is correct too.